### PR TITLE
[3.x] OIDC multi-tenant and lazy loading implementation

### LINKED ISSUE #5515: [3.x] OIDC multitenant support
Backport of #4936 for Helidon 3.x

### LINKED ISSUE #5518: [3.x] OIDC lazy loading of metadata
Backport of #4383 for Helidon 3.x

### PR DESCRIPTION
**Pull request by Verdent**
Fixes #5515
Fixes #5518

The only change which was required to do (in comparison to the 2.x version), was to add `common-io` as test dependency. There was an issue with `resteasy-jaxrs` and I had to exclude it. That introduced this need to add `common-io`.

### CODE REVIEW

File: security/providers/oidc-common/src/main/java/io/helidon/security/providers/oidc/common/BaseBuilder.java [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+abstract class BaseBuilder<B extends BaseBuilder<B, T>, T extends TenantConfig> implements Builder<B, T> {
+
+    static final String DEFAULT_SERVER_TYPE = "@default";
+    static final String DEFAULT_BASE_SCOPES = "openid";
+    static final String DEFAULT_REALM = "helidon";
+    static final boolean DEFAULT_JWT_VALIDATE_JWK = true;
+    static final int DEFAULT_TIMEOUT_SECONDS = 30;
+    private static final JsonReaderFactory JSON = Json.createReaderFactory(Collections.emptyMap());
+
+    @SuppressWarnings("unchecked")
```
Comment (tomas-langer) [inline security/providers/oidc-common/src/main/java/io/helidon/security/providers/oidc/common/BaseBuilder.java:46]: this is no longer needed. Return `identity()` instead (supported by Builder in common)

File: security/providers/oidc-common/src/main/java/io/helidon/security/providers/oidc/common/BaseBuilder.java [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+        config.get("validate-with-jwk").asBoolean().ifPresent(this::validateJwtWithJwk);
+        config.get("issuer").asString().ifPresent(this::issuer);
+        config.get("audience").asString().ifPresent(this::audience);
+
+        // type of the identity server
+        // now uses hardcoded switch - should change to service loader eventually
+        config.get("server-type").asString().ifPresent(this::serverType);
+
+        config.get("client-timeout-millis").asLong().ifPresent(this::clientTimeoutMillis);
+        return me;
```
Comment (tomas-langer) [inline security/providers/oidc-common/src/main/java/io/helidon/security/providers/oidc/common/BaseBuilder.java:121]: replace all of these with `return identity();`

File: security/providers/oidc-common/src/main/java/io/helidon/security/providers/oidc/common/Tenant.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+import io.helidon.webclient.security.WebClientSecurity;
+
+import jakarta.json.JsonObject;
+import jakarta.ws.rs.client.Client;
+import jakarta.ws.rs.client.ClientBuilder;
+import jakarta.ws.rs.client.WebTarget;
+import org.glassfish.jersey.client.authentication.HttpAuthenticationFeature;
+
+/**
+ * Holder of the tenant configuration resolved in runtime. Used for OIDC lazy loading.
```
Comment (tomas-langer) [inline security/providers/oidc-common/src/main/java/io/helidon/security/providers/oidc/common/Tenant.java:38]: `resolved at runtime`

File: security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcSupport.java [hunk 4]
```diff
@@ -242,6 +268,48 @@ private void processLogout(ServerRequest req, ServerResponse res) {
...
+        } else {
+            if (LOGGER.isLoggable(Level.FINEST)) {
+                LOGGER.finest("Missing tenant id, could not find in either of: " + missingLocations
+                                      + "Falling back to the default tenant id: " + DEFAULT_TENANT_ID);
+            }
+            return Single.just(DEFAULT_TENANT_ID);
+        }
+    }
+
+    private Tenant obtainCurrentTenant(String tenantName) {
```
Comment (tomas-langer) [inline security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcSupport.java:301]: This method is blocking (`Tenant.create()` may invoke HTTP requests against metadata endpoint). This cannot run in web server request as designed here

File: security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcSupport.java [hunk 5]
```diff
@@ -274,24 +342,27 @@ private void processOidcRedirect(ServerRequest req, ServerResponse res) {
     }
 
     private void processCode(String code, ServerRequest req, ServerResponse res) {
-        WebClient webClient = oidcConfig.appWebClient();
+        String tenantName = req.queryParams().first(oidcConfig.tenantParamName()).orElse(TenantConfigFinder.DEFAULT_TENANT_ID);
+
+        Tenant tenant = obtainCurrentTenant(tenantName);
```
Comment (tomas-langer) [inline security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcSupport.java:347]: `obtainCurrentTenant` may be a blocking operation

File: security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcUtil.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import io.helidon.common.http.FormParams;
+import io.helidon.security.providers.oidc.common.OidcConfig;
+import io.helidon.security.providers.oidc.common.TenantConfig;
+
+class OidcUtil {
+
+    private OidcUtil() {
+    }
+
+    static void updateRequest(OidcConfig.RequestType type, TenantConfig tenantConfig, FormParams.Builder form) {
```
Comment (tomas-langer) [inline security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcUtil.java:28]: Why is this in a utility class? How many classes use this?

File: security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/TenantAuthenticationHandler.java [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+                                                                                    + ", entity: " + message)),
+                                        (t, message) ->
+                                                Optional.of(collector.fatal(t,
+                                                                            "Failed to validate token, request failed: "
+                                                                                    + message)));
+            };
+        }
+        // clean the scope audience - must end with / if exists
+        String configuredScopeAudience = tenantConfig.scopeAudience();
+        if (null == configuredScopeAudience || configuredScopeAudience.isEmpty()) {
```
Comment (tomas-langer) [inline security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/TenantAuthenticationHandler.java:152]: flip condition please

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcUtil.java (hunk 6 above)
Comment (Verdent) [reply security/providers/oidc/src/main/java/io/helidon/security/providers/oidc/OidcUtil.java:28]: Not an issue (as discussed over the Zoom)

**Review by tomas-langer — APPROVED** [review]

**Pull request merged**